app/gui/view/graph-editor/src/component/node/input/widget.rs

```diff
@@ -8,54 +8,181 @@ use ensogl::data::color;
 use ensogl::display;
 use ensogl::display::object::event;
 use ensogl_component::drop_down::Dropdown;
-use frp::stream::EventEmitter;
 
 
 
-// ==================
-// === NodeWidget ===
-// ==================
+/// =================
+/// === Constants ===
+/// =================
+
+const DOT_COLOR: color::Lch = color::Lch::new(0.56708, 0.23249, 0.71372);
+
+
+
+// ===========
+// === FRP ===
+// ===========
 
 ensogl::define_endpoints_2! {
     Input {
-        set_current_value(Option<ImString>),
-        set_focused(bool),
+        set_metadata   (Option<Metadata>),
+        set_node_data  (NodeData),
+        set_current_value (Option<ImString>),
+        set_focused       (bool),
+        set_visible       (bool),
     }
     Output {
         value_changed(Option<ImString>),
     }
 }
 
-/// Possible widgets for a node input.
-///
-/// Currently all widget types are hardcoded. This is likely to be a temporary solution. In the
-/// future the set of widget types might be dynamic, similar to visualizations.
-#[derive(Clone, Debug, CloneRef)]
-pub enum Widget {
-    /// A widget for selecting a single value from a list of available options.
-    SingleChoice(SingleChoice),
+/// Widget metadata that needs to be fetched from the graph asynchronously.
+#[derive(Debug, Clone)]
+pub struct Metadata {
+    /// The kind of widget to use.
+    pub kind:            Kind,
+    /// The widget display mode.
+    pub display:         Display,
+    /// Dynamically fetched widget entries.
+    pub dynamic_entries: Vec<ImString>,
+}
+
+/// Widget display mode. Determines when the widget should be expanded.
+#[derive(serde::Deserialize, Debug, Clone, Copy)]
+pub enum Display {
+    /// The widget should always be in its expanded mode.
+    Always,
+    /// The widget should only be in its expanded mode when it has non-default value.
+    #[serde(rename = "When_Modified")]
+    WhenModified,
+    /// The widget should only be in its expanded mode whe the whole node is expanded.
+    #[serde(rename = "Expanded_Only")]
+    ExpandedOnly,
+}
+
+impl Default for Display {
+    fn default() -> Self {
+        Self::Always
+    }
+}
+
+/// The data of node port that this widget is attached to. Available immediately after widget
+/// creation. Can be updated later when the node data changes.
+#[derive(Debug, Clone, Default)]
+pub struct NodeData {
+    /// Argument info of the node port that this widget is attached to.
+    pub argument_info: span_tree::ArgumentInfo,
+    /// The current height of the node that the widget can use for layout.
```

Comment:
again, these docs are obvious. Having them makes the code longer to read. Lets use docs which explain non-bovious things only.